Code/Framework/AzCore/AzCore/Debug/Trace.h

```diff
@@ -39,14 +37,122 @@ namespace AZ
+                return m_printCallstack;
+            }
+            O3DEKERNEL_API void SetAlwaysPrintCallstack(bool enable)
+            {
+                m_printCallstack = enable;
+            }
+
+            O3DEKERNEL_API virtual void Assert(const char* fileName, int line, const char* funcName, const char* format, ...)
+            {
+                char message[s_maxMessageLength];
+                va_list mark;
+                va_start(mark, format);
+                azvsnprintf(message, s_maxMessageLength, format, mark);
+                va_end(mark);
+                (void)fprintf(stderr, "Assert: %s:%d (%s): %s\n", fileName, line, funcName, message);
+            }
+            O3DEKERNEL_API virtual void Error(const char* fileName, int line, const char* funcName, const char* window, const char* format, ...)
+            {
+                char message[s_maxMessageLength];
+                va_list mark;
+                va_start(mark, format);
+                azvsnprintf(message, s_maxMessageLength, format, mark);
+                va_end(mark);
+                (void)fprintf(stderr, "Error: %s:%d (%s): %s: %s\n", fileName, line, funcName, window, message);
+            }
+            O3DEKERNEL_API virtual void Warning(const char* fileName, int line, const char* funcName, const char* window, const char* format, ...)
+            {
+                char message[s_maxMessageLength];
+                va_list mark;
+                va_start(mark, format);
+                azvsnprintf(message, s_maxMessageLength, format, mark);
+                va_end(mark);
+                (void)fprintf(stderr, "Warning: %s:%d (%s): %s: %s\n", fileName, line, funcName, window, message);
+            }
+            O3DEKERNEL_API virtual void Printf(const char* window, const char* format, ...)
+            {
+                char message[s_maxMessageLength];
+                va_list mark;
+                va_start(mark, format);
+                azvsnprintf(message, s_maxMessageLength, format, mark);
+                va_end(mark);
+                (void)fprintf(stderr, "%s: %s\n", window, message);
+            }
+            O3DEKERNEL_API virtual void Output(const char* window, const char* message)
+            {
+                (void)fprintf(stderr, "%s: %s\n", window, message);
+            }
+            O3DEKERNEL_API virtual void RawOutput(const char* window, const char* message)
+            {
+                (void)fprintf(stderr, "%s: %s\n", window, message);
+            }
+            O3DEKERNEL_API virtual void PrintCallstack(const char* /*window*/, unsigned int /*suppressCount*/ = 0, void* /*nativeContext*/ = nullptr) {}
+
+        private:
+            inline static constexpr size_t s_maxMessageLength = 4096;
+            O3DEKERNEL_API inline static ITrace* tracer{};
+
+            LogLevel m_logLevel = LogLevel::Info;
+            bool m_printCallstack = false;
+        };
+
+        class Trace
+            : public ITrace
```

Comment:
Is `Trace` the one-and-only implementer of `ITrace`?  Are we leaving the door open for other classes to implement that interface?

Another way to look at this, assuming there's only really one implementation, is to move all concrete implementations to the `ITrace` class and make it non-virtual, then the `AZ::Debug::Trace` class is a thin layer of static functions that simply call `ITrace::Instance()` and forward the call.  That way all the `AZ::Debug::Trace` calls wouldn't need to be modified and the calls to get the `Instance()` are hidden directly behind them.

Reply:
`Trace`'s implementation cannot be the same thing as `ITrace`'s. `ITrace`'s symbols are defined in `O3DEKernel`. `Trace`'s symbols are defined in `AzCore`. `Trace` depends on `EBus` and `Console`, which are not present in `O3DEKernel`.

Reply:
I see, thanks for clarifying.